Added access check for `getUserById` and `getUserByEmail`

Slartibartfass2 opened the pull request:
Closes #54 

## What I have made

Now, only

- the server admin,
- the requested user itself, or
- a user who is in a project together with the requested user

can access a user by its ID or Email.

This required implementing some repository methods in `ProjectMemberTableRepo`.

I also changed the code so that the repo layer only uses UUIDs and not strings as IDs because always converting the strings to UUIDs was getting cumbersome and ugly. As the dummy user will be replaced soon, it will be less of a problem in the service layer.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add `getProjectMembersInSameProjectsAsUser` for user access check

Slartibartfass2 committed:
refactor: store strings instead of composite IDs

For testing, it is impossible to create composite IDs

Slartibartfass2 committed:
refactor: set dummyUserId before each test

This enables modifying the dummy user ID for specific test cases.

Slartibartfass2 committed:
feat: add access check for `getUserById`

Slartibartfass2 committed:
test: extend tests for `getUserById` by testing the access check

Slartibartfass2 committed:
feat: add access check for `getUserByEmail`

Slartibartfass2 committed:
test: extend tests for `getUserByEmail` by testing access check

Slartibartfass2 committed:
refactor: let repo layer use actual UUIDs instead of string IDs

This prevents the whole toString conversations and moves the conversion to the service layer where it will be replaced when the user management is established.

Slartibartfass2 committed:
test: add TestSpecificException to ensure correct exception mocking

domienderle commented on src/test/kotlin/service/user/GetUserByEmailTest.kt:
```diff
@@ -7,30 +7,113 @@ import org.junit.jupiter.api.Test
+            val requestingUser = DataBuilder.createExampleUser(id = UUID.fromString(dummyUserId))
+            coEvery { userRepoMock.getUserById(dummyUserId!!) } returns requestingUser
+            coEvery { userRepoMock.getUserByEmail(requestEmail) } returns requestingUser
+            coEvery { projectMemberRepoMock.getProjectMembersInSameProjectsAsUser(any()) } returns listOf()
```
Is this line really necessary, or couln't it just be removed? As I tested it, it doens't make any difference.

Slartibartfass2 replied:
No, the call isn't made. The logic was structured differently before, but now we can remove this.

domienderle commented on src/main/kotlin/model/SnowballRException.kt:
```diff
@@ -83,8 +87,15 @@ sealed class SnowballRException(
         sealed class Single(
             currentUserId: String,
             accessedEntityType: String,
+            identifier: String? = "ID",
```
Just a small thing, but I would propose to sort the parameters, like this: `currentUserId`, `accessedEntityId`, `accessedEntityType`, `identifier`, or at least, that the `accessedEntity` parameters stand together.

domienderle commented on src/main/kotlin/repository/association/ProjectMemberTableRepo.kt:
```diff
@@ -0,0 +1,121 @@
+                }
+
+        ProjectMemberTable
+            .selectAll()
```
Is the `selectAll()` really necessary, when you call `where` in the next call? Can't you just use `select`? Applies to all of the methods in this class.

Slartibartfass2 replied:
Yes, it is necessary to retrieve all columns via `selectAll` and then filter according to the ID using `where`. If we used `select`, we would have to pass every column that we need for this expression, and I don't think we require listing all columns because we need them all if we want to map the result row to an entity object later.

domienderle replied:
Ah ok. Yeah this sounds like it makes sense. Thanks for the explanation.

domienderle commented on src/main/kotlin/service/UserService.kt:
```diff
@@ -33,28 +37,75 @@ interface IUserService {
+        if (requestedUserId == currentUser.id) return
+
+        // Check whether requesting user is in a same project as the requested user
+        val isInSameProject =
```
Does this call makes sense? Isn't it always true, because you request the project members of the project, where the `currentUser` is participant. After that you check, whether the `currentUser` is part of this list, which is true in every case, right?

Slartibartfass2 replied:
Yes, you're right. That should've been the requested user. Great catch!

domienderle commented on src/main/kotlin/service/UserService.kt:
```diff
@@ -33,28 +37,75 @@ interface IUserService {
+
+        val isRequestedUser = requestingUserId == currentUser.id
+
+        // Don't rerequest the user if it is the current user itself
```
```suggestion
        // Don't re-request the user if it is the current user itself
```

domienderle commented on src/test/kotlin/repository/association/ProjectMemberTableRepoTest.kt:
Use the `setupProject` method consequently. In some tests you use it and in other tests you call this manually:
```kotlin
val project = createExampleProject()
val member = repo.addUserToProject(testUserId, project.id)
```
I would even suggest to add a extra parameter to the method like this: 
```kotlin
suspend fun setupProject(vararg additionalUsers: Int): Project {
    val project = createExampleProject()
    repo.addUserToProject(testUserId, project.id)
    additionalUsers.forEach { addTestMember(it, project.id) }
    return project
}
```

Slartibartfass2 replied:
Great suggestion. I changed it wherever it was applicable.

domienderle replied:
Now it looks really good.

domienderle commented on src/test/kotlin/service/user/GetUserByIdTest.kt:
```diff
@@ -3,35 +3,152 @@ package se.uulm.snowballr.backend.service.user
+        }
+
+    @Test
+    fun `When the requesting the current user fails, then an exception is thrown`() =
```
```suggestion
    fun `When requesting the current user fails, then an exception is thrown`() =
```
or
```suggestion
    fun `When the request for the current user fails, then an exception is thrown`() =
```

domienderle reviewed (changes requested):
Overall good implementation. A couple of smaller things need a rework.

Slartibartfass2 committed:
refactor: remove duplicated 'project' in ProjectMemberTableRepo methods

Renamed `getProjectMembersOfProject` to `getMembersOfProject` and `getProjectMembersInSameProjectsAsUser` to `getMembersInSameProjectsAsUser`.

Slartibartfass2 committed:
fix: request the members in the same projects of the requested user not the requesting one

Slartibartfass2 committed:
test: use test helper consistently

Slartibartfass2 committed:
chore: fix comments and method names

Slartibartfass2 committed:
refactor: reorder parameter list of exception

Slartibartfass2 reviewed (commented):
Thank you for the review. I will rebase as soon as we're done with the discussions.

Slartibartfass2 committed:
feat: add filters for user retrieval calls

Only active and active unconfirmed users can be retrieved.

domienderle commented:
Now it looks really good. You can merge it now.

Slartibartfass2 committed:
style: format tests after rebase
